### PR TITLE
fix(#187): Remove hidden-test leakage from cron_next_fire and iban_validator prompts

### LINKED ISSUE #187: Remove hidden-test leakage from cron_next_fire and iban_validator prompts
## Context

Follow-up to #173. Two task prompts leak ~35–40% of their hidden test set inline as worked examples, undermining the hidden/visible split that gives these tasks discriminating power.

## Findings to address

- **`cron_next_fire`**: 7/20 hidden tests (35%) appear verbatim in the prompt examples — same expressions and same \`after\` anchor (\`2024-01-01 12:00\`).
- **`iban_validator`**: 10/25 hidden tests (40%) appear verbatim in the prompt examples — canonical DE/GB/FR/BE IBANs plus several negative cases.

### 1. \`cron_next_fire/prompt.md\`
Shift the \`after\` anchor in the example table from \`2024-01-01 12:00\` to a date no hidden test uses (e.g. \`2025-03-15 09:30\`) and recompute the 7 next-fire values. Same illustrative coverage, zero overlap with \`_TESTS\`.

### 2. \`iban_validator/prompt.md\`
Replace the example IBANs with inputs not present in \`grader/hidden.py:_TESTS\`. Negative-case examples (bad checksum, wrong length, unknown country, empty, None) stay structurally identical but use different literal values.

### 3. CI guard — \`tests/test_prompt_hidden_disjoint.py\`
Walk every \`problems/artifact/**/\` directory, extract literal inputs from \`grader/hidden.py:_TESTS\`, scan the corresponding \`prompt.md\` for any occurrence, and fail on intersection. Prevents this class of leak for all future tasks.

### 4. Verify
Run \`python -m swebench artifact verify\` on both tasks after edits to confirm graders still match reference outputs.

## Definition of done

- [ ] Hidden test sets for both tasks are disjoint from prompt examples (zero overlap).
- [ ] CI check added that flags prompt/hidden overlap for any artifact task, so this class of leak is prevented for future tasks.
- [ ] \`artifact verify\` passes for both tasks.

### PR DESCRIPTION
**Pull request by hyang0129**
Closes #187

## What changed

Two artifact task prompts leaked portions of their hidden test inputs verbatim as worked examples, undermining benchmark discriminating power. An agent that simply pattern-matched the prompt examples against expected outputs could score partial credit on the hidden test set without actually implementing the algorithm. This PR fixes both affected tasks and adds a CI guard to prevent this class of leak for all future tasks.

### `cron_next_fire/prompt.md` — anchor shift

The original prompt examples used `after = datetime(2024, 1, 1, 12, 0, 0)` as the reference instant. The hidden test `_TESTS` uses this exact anchor for 7 of its 20 cases, pairing it with expressions like `* * * * *`, `0 9-17 * * 1-5`, etc. Any agent that saw the worked examples could determine the expected output for those test cases directly.

**Fix:** The anchor was shifted to `2025-03-15 09:30:00` (a Saturday morning), which appears in no hidden test. All 7 example next-fire values were recomputed against the new anchor. The cron expressions themselves are retained — they are necessary vocabulary for explaining the cron grammar, not uniquely identifying data.

### `iban_validator/prompt.md` — example IBAN replacement

The original prompt showed `DE89 3704 0044 0532 0130 00`, `GB82WEST12345698765432`, `FR1420041010050500013M02606`, and `BE68539007547034` as positive examples. All four appear verbatim in `_TESTS`, and the negative example `DE89370400440532013001` (wrong checksum by one digit) also matches a hidden test. An agent with the prompt could answer those hidden tests by lookup rather than by implementing the validator.

**Fix:** All example IBANs were replaced with mod-97-valid IBANs from supported countries that are not present in `_TESTS`:
- `DE77200400600532013001` — Deutsche Bank format, different account
- `GB29NWBK60161331926819` — NatWest format  
- `CH5604835012345678009` — Swiss format
- `BE15430609679000` — Belgian format

Negative cases were updated to use the new valid examples as the base (wrong checksum, wrong length, unknown country `YY`) so the structural variety is preserved with different literal values.

### `tests/test_prompt_hidden_disjoint.py` — CI guard (new file)

The new test walks every artifact task directory under `problems/artifact/**/`, extracts all string literals from the top-level `_TESTS` variable in `grader/hidden.py` via AST parsing (no import, no execution), and fails if any of those strings appear verbatim in `prompt.md`.

**Design decisions:**
- AST parsing (`ast.AnnAssign` and `ast.Assign` for `_TESTS`) is safe and offline; it handles both plain and type-annotated assignments.
- Case-insensitive comparison so lowercase IBAN inputs also match.
- Strings shorter than 4 characters are excluded (country codes, keywords are expected in specs).
- A `_PROMPT_LEAK_EXEMPT` dict documents per-task exemptions for inputs that are inherent syntax tokens (cron expressions, semver version strings, ISO duration patterns, arithmetic expressions). These exemptions are documented with rationale comments and are narrow — only the exact listed strings are skipped.

The test is parametrised per task (48 tasks, all passing) so CI output names the specific task and leaked string on failure.

### How components interact

**Before**: Author writes a task → no validation that prompt examples are disjoint from hidden tests → leakage goes undetected until a human review catches it.

**After**: Author writes a task → `pytest tests/test_prompt_hidden_disjoint.py` runs in CI → any string in `_TESTS` that appears in `prompt.md` (and is not in `_PROMPT_LEAK_EXEMPT`) causes immediate failure → author fixes the prompt before merging.

### Default execution path

**Before (broken):** `cron_next_fire` prompt shows `(* * * * *, 2024-01-01 12:00) → 2024-01-01 12:01`. Hidden test #1 is exactly `("* * * * *", datetime(2024, 1, 1, 12, 0), datetime(2024, 1, 1, 12, 1))`. An agent looking at the examples could predict this output without implementing `next_fire`.

**After (fixed):** `cron_next_fire` prompt shows `(* * * * *, 2025-03-15 09:30) → 2025-03-15 09:31`. Hidden test #1 uses anchor `datetime(2024, 1, 1, 12, 0)` — completely different. The agent must implement the algorithm to get correct answers.

### Edge cases and error handling

The CI guard handles:
- Tasks without string inputs in `_TESTS` (e.g. purely numeric inputs) — skipped gracefully.
- Tasks with invalid Python in `hidden.py` — `SyntaxError` caught, returns empty set.
- Tasks with either `ast.Assign` or `ast.AnnAssign` for `_TESTS` — both forms handled.
- Tasks where exempt strings are a complete superset of test strings — test returns early.

### What was intentionally not changed

- `grader/hidden.py` for both tasks: no changes to the hidden test sets.
- `grader/reference_output.*` for both tasks: no changes (verified by `tools/verify_graders.py`).
- Workspace files for both tasks: unchanged.
- The cron expressions themselves (`* * * * *`, `0 9-17 * * 1-5`, etc.) remain in the prompt as necessary grammar illustrations — they are exempted in `_PROMPT_LEAK_EXEMPT`.

## Tier / approach

Tier 2 — Planner → Coder A (prompt fixes) + Tester (CI guard) → validation

## Acceptance criteria

- [x] Hidden test sets for both tasks are disjoint from prompt examples (zero overlap verified by `test_prompt_hidden_disjoint`).
- [x] CI check added that flags prompt/hidden overlap for any artifact task (48 tasks, all passing).
- [x] `artifact verify` passes for both tasks (`verification_heavy__cron_next_fire`: all 20 cases passed; `verification_heavy__iban_validator`: all 25 cases passed).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by hyang0129 — COMMENTED** [review]
# Review Findings — PR #194 (Cycle 0 / Initial Review)

**PR**: fix(#187): Remove hidden-test leakage from cron_next_fire and iban_validator prompts
**Branch**: fix/issue-187-prompt-hidden-leakage → main
**Reviewer model**: claude-opus-4-7 (default)

Summary: 0 `fix`-tier findings, 1 `propose`-tier finding, 0 `escalate`-tier findings.

**Zero critical/major tier `fix` findings** — the fix loop will exit early. The single tier `propose` finding is surfaced as a PR comment for author review.

---

## Tier: `fix`

(None)

---

## Tier: `propose`

### [Tier: propose] [Severity: minor] Unused `import re` in test file

**ID**: F-0-1
**File**: `tests/test_prompt_hidden_disjoint.py:33`
**Problem**: `import re` appears on line 33 but the `re` module is never referenced anywhere in the file. The implementation uses only `ast`, `pathlib.Path`, `typing.Iterator`, and `pytest`. This is a dead import — likely a draft residue from an earlier regex-based scanning approach before the current case-insensitive `.upper()` + `in` strategy was chosen. Will trigger standard linters (flake8 F401, ruff F401).
**Tier rationale**: Removing the import produces exactly one correct outcome (deletion of one line), which sounds like tier `fix`. However, the sniff test applies: the problem reads "could be cleaner / lint noise," not "will crash / will return wrong value." Dead imports are style-tier. Tier `propose` is correct.
**Proposed diff**:
```diff
--- a/tests/test_prompt_hidden_disjoint.py
+++ b/tests/test_prompt_hidden_disjoint.py
@@ -30,7 +30,6 @@ from __future__ import annotations
 
 import ast
-import re
 from pathlib import Path
 from typing import Iterator
```
**Rationale**: Semantically inert; eliminates lint noise and removes a misleading hint about the module's approach.
**Decision required**: no
**Parallelizable**: yes
**Conflicts with**: none

---

## Tier: `escalate`

(None)

---

## Overall Assessment

The PR correctly and completely addresses the reported prompt/hidden-test leakage in three components:

1. **`cron_next_fire/prompt.md`** — Anchor shift from `2024-01-01 12:00:00` to `2025-03-15 09:30:00` (Saturday) is well-chosen. All 7 recomputed example next-fire values verified arithmetically correct against the new anchor. The hidden tests continue to use the `2024-01-01 12:00:00` anchor — disjoint. Cron expressions correctly remain in the prompt and are properly exempted in `_PROMPT_LEAK_EXEMPT`.

2. **`iban_validator/prompt.md`** — All 5 originally-leaked IBANs replaced with mod-97-valid alternatives from the same supported country set. Structural correctness verified: DE (22 chars), GB (22 chars), CH (21 chars), BE (16 chars) all match the supported-country length table. All positive examples confirm `int(rearranged) % 97 == 1`. Wrong-checksum and wrong-length negatives use the new DE77... base, preserving structural pedagogical variety.

3. **`tests/test_prompt_hidden_disjoint.py`** — CI guard is well-designed: AST-only extraction (no import/exec), parametrized per task, case-insensitive comparison, narrowly-scoped `_PROMPT_LEAK_EXEMPT` entries with documented rationale comments. All 438 non-integration tests pass (48/48 disjoint-check tasks pass). No bugs, no logic errors, no security issues, no missing error handling.

**Comment by hyang0129** [comment]
## Proposed changes (not auto-applied) and escalations

These findings were identified by the Reviewer but fall outside the auto-fix scope. The fix loop will not apply them. Please review and decide.

### Proposed changes — diffs for your consideration

#### F-0-1 [minor] Unused `import re` in test file
**File**: `tests/test_prompt_hidden_disjoint.py:33`

`import re` is present on line 33 but the `re` module is never used. This is likely a draft residue from an earlier regex-based approach. Will trigger flake8/ruff F401.

```diff
--- a/tests/test_prompt_hidden_disjoint.py
+++ b/tests/test_prompt_hidden_disjoint.py
@@ -30,7 +30,6 @@ from __future__ import annotations
 
 import ast
-import re
 from pathlib import Path
 from typing import Iterator
```

### Escalated — candidates for follow-up issues

None.

**Comment by hyang0129** [comment]
## Automated Review-Fix Summary

**Cycles run**: exited early at cycle 1 (no critical/major tier `fix` findings)
**Commits added**: 0
**Reviewer model**: claude-opus-4-7 (default)

### Final review state
- Critical tier `fix` findings remaining: 0
- Major tier `fix` findings remaining: 0
- Minor tier `fix` findings remaining: 0
- Tier `propose` findings surfaced: 1 (see proposals comment)
- Tier `escalate` findings surfaced: 0
- Overall: CLEAN — no critical/major tier `fix` remaining

### Tier `fix` findings applied (0)

None — no tier `fix` findings were identified.

### Tier `propose` findings (not auto-applied — author decides)

| ID | Severity | Title | File |
|---|---|---|---|
| F-0-1 | minor | Unused `import re` in test file | `tests/test_prompt_hidden_disjoint.py:33` |

See proposals comment above for the one-line removal diff.

### Tier `escalate` findings (candidates for follow-up issues)

None.

### Outstanding tier `fix` findings (not applied)

None — all tier `fix` findings resolved (none were found).

### Intent validation

No automated fixes were applied during the review-fix cycle. The pre-loop and
post-loop states are identical (pre-loop SHA: af7e062cae4a6947436b9ad576ef1300e0ec23f3).

The fix loop exited early at cycle 1 because the Reviewer found zero critical or major
tier `fix` findings. No Fixer agents ran. Therefore there are no automated changes
to validate against the original PR intent.

No intent risks detected — all automated fixes are consistent with the original PR intent.

---
